### PR TITLE
Fix index dimension mismatch in `TensorNetwork` construction

### LINKED ISSUE #155: `Quantum` constructor does not error for inconsistent dimensions in tensor networks
Currently, it is possible to create a `Chain` ansatz that contains tensors with inconsistent dimension sizes between tensors. In my opinion, there should be an error if one tries to create a `Quantum` tensor network with those inconsistencies.
I suggest using the new function `Tenet.__check_index_sizes` before returning the `Quantum` tensor network, and if this function is false, then `throw` an error.

### Example
```julia
julia> using Qrochet

julia> mps = Chain(State(), Open(), [rand(2, 3), rand(2, 6, 4), rand(2, 5)]) # Does not error
MPS (inputs=0, outputs=3)

julia> inds.(tensors(mps))
3-element Vector{Vector{Symbol}}:
 [:F, :I]
 [:G, :I, :J]
 [:H, :J]

julia> size.(tensors(mps))
3-element Vector{Tuple{Int64, Int64, Vararg{Int64}}}:
 (2, 3)
 (2, 6, 4)
 (2, 5)
```

### PR DESCRIPTION
**Pull request by Todorbsc**
Closes #155

### CODE REVIEW

**Comment by Todorbsc** [comment]
@jofrevalles @mofeing the dimension mismatch seems to be fixed with the verification I implemented in `TensorNetwork` construction. I "uncommented" the TN dimension mismatch construction test and now it raises the proper exception as well as I tested the `Chain` example that Jofre commented in the [issue](https://github.com/bsc-quantic/Tenet.jl/issues/155#issue-2362798480), what raises an error now:
```julia
julia> mps = Chain(State(), Open(), [rand(2, 3), rand(2, 6, 4), rand(2, 5)])
ERROR: DimensionMismatch: Index D has inconsistent dimension: [3, 6]
Stacktrace:
 [1] TensorNetwork(tensors::Vector{Tensor{Float64, N, A} where {N, A<:AbstractArray{Float64, N}}})
   @ Tenet ~/QuanticGit/Tenet.jl/src/TensorNetwork.jl:58
 [2] Chain(::State, boundary::Open, arrays::Vector{Array{Float64}}; order::Tuple{Symbol, Symbol, Symbol})
   @ Tenet ~/QuanticGit/Tenet.jl/src/Ansatz/Chain.jl:99
 [3] Chain(::State, boundary::Open, arrays::Vector{Array{Float64}})
   @ Tenet ~/QuanticGit/Tenet.jl/src/Ansatz/Chain.jl:63
 [4] top-level scope
   @ REPL[3]:1
```
Also, let me know if there's something else needed, specially with what you commented in the issue about the `unsafe_region`.

**Review by jofrevalles — APPROVED** [review]
Okay, nice! Since we already have tests for `@unsafe_region`, and these are passing, we can merge this!

**Comment by mofeing** [comment]
Do you mind benchmarking with and without safe guards?

File: src/TensorNetwork.jl [hunk 1]
```diff
@@ -45,14 +45,19 @@ struct TensorNetwork <: AbstractTensorNetwork
...
             for index in inds(tensor)
                 # TODO use lambda? `Tensor[]` might be reused
                 push!(get!(dict, index, Tensor[]), tensor)
             end
             dict
         end
 
+        # Check for inconsistent index dimensions
+        for ind in keys(indexmap)
+            dims = map(tns -> size(tns)[findfirst(==(ind), tensormap[tns])], indexmap[ind])
```
Comment (mofeing) [inline src/TensorNetwork.jl:57]:
```suggestion
            dims = map(tensor -> size(tensor, ind), indexmap[ind])
```

**Review by mofeing — CHANGES_REQUESTED** [review]

**Comment by Todorbsc** [comment]
> Do you mind benchmarking with and without the safe guards / checks for some large TNs? Like 100-1000 tensors

I benchmarked for `rand(TensorNetwork, 1000, 3)`.

Without the indices safe region:
```julia
julia> @benchmark tn = rand(TensorNetwork, 1000, 3)
BenchmarkTools.Trial: 88 samples with 1 evaluation.
 Range (min … max):  17.460 ms … 560.275 ms  ┊ GC (min … max):  0.00% … 12.05%
 Time  (median):     25.241 ms               ┊ GC (median):     2.20%
 Time  (mean ± σ):   56.874 ms ±  82.945 ms  ┊ GC (mean ± σ):  22.67% ± 24.72%

  █▅▁                                                           
  █████▅▁▅▁▁▅▆▆▅▇▅▁▁▅▁▆▁▁▁▁▁▅▁▁▁▁▁▁▁▁▁▅▁▁▁▅▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▅ ▁
  17.5 ms       Histogram: log(frequency) by time       407 ms <

 Memory estimate: 25.09 MiB, allocs estimate: 64188.
 ```
 
 With the indices safe region:
 ```julia
julia> @benchmark tn = rand(TensorNetwork, 1000, 3)
BenchmarkTools.Trial: 89 samples with 1 evaluation.
 Range (min … max):  23.384 ms … 505.261 ms  ┊ GC (min … max):  0.00% … 15.44%
 Time  (median):     30.132 ms               ┊ GC (median):     0.00%
 Time  (mean ± σ):   61.098 ms ±  85.135 ms  ┊ GC (mean ± σ):  12.13% ± 14.75%

  █▃ ▁                                                          
  ████▇█▁▅▁▇▁▁▁▆▁▁▆▅▁▁▁▁▁▁▁▁▁▁▁▅▅▁▁▁▁▁▁▁▁▁▁▅▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▅ ▁
  23.4 ms       Histogram: log(frequency) by time       485 ms <

 Memory estimate: 28.02 MiB, allocs estimate: 103174.
 ```

**Comment by Todorbsc** [comment]
@mofeing Another benchmark for `rand(TensorNetwork, 200, 3)`.

Using checks:
```julia
julia> @benchmark tn = rand(TensorNetwork, 200, 3)
BenchmarkTools.Trial: 400 samples with 1 evaluation.
 Range (min … max):   2.218 ms … 253.500 ms  ┊ GC (min … max):  0.00% … 23.87%
 Time  (median):      3.163 ms               ┊ GC (median):     0.00%
 Time  (mean ± σ):   12.499 ms ±  28.164 ms  ┊ GC (mean ± σ):  31.62% ± 21.23%

  █▄▁                                                           
  ███▇▅▆▅▅▅▄▅▅▅▅▄▅▁▅▁▁▄▄▁▁▁▄▁▁▁▄▄▆▁▁▅▆▁▅▁▄▁▄▄▁▄▁▅▁▁▁▁▁▁▁▄▁▄▁▄▄ ▆
  2.22 ms       Histogram: log(frequency) by time       129 ms <

 Memory estimate: 2.80 MiB, allocs estimate: 19763.
 ```
 
 Wiithout checks:
 ```julia
julia> @benchmark tn = rand(TensorNetwork, 200, 3)
BenchmarkTools.Trial: 353 samples with 1 evaluation.
 Range (min … max):   1.240 ms …   1.031 s  ┊ GC (min … max):  0.00% …  0.32%
 Time  (median):      2.028 ms              ┊ GC (median):     0.00%
 Time  (mean ± σ):   14.150 ms ± 65.195 ms  ┊ GC (mean ± σ):  30.34% ± 22.77%

  █                                                            
  ██▅▆▄▅▅▁▁▄▁▁▁▁▄▅▅▅▄▆▄▆▁▁▁▄▁▄▁▁▁▁▄▁▁▁▁▁▄▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▄ ▅
  1.24 ms      Histogram: log(frequency) by time       222 ms <

 Memory estimate: 2.84 MiB, allocs estimate: 11968.
 ```
 
 Min. time: Checks `2.218 ms` vs. No checks `1.240 ms`.

**Comment by mofeing** [comment]
Thanks!

> Correct me if I am wrong, but as regularity may influence the times, I would just look at the minimum time: Without the checks 17.460 ms vs. With the checks 23.384 ms.

Well, the min time can also be affected by jitter so I would take the mean which means there is a 20% increase in runtime... but it's still 25ms which is ok.

**Review by mofeing — APPROVED** [review]

**Pull request merged**